### PR TITLE
Full parametrized decorator

### PR DESCRIPTION
**Pull request by elijahbenizzy**
Consolidates logic behind all `parametrized` decorators

## Changes

- adds `full_parametrized` decorator
- changes others to use it
- Adds a smoke_screen_test with a common set of decorators that we use
- Adds more testing for distributed integrations

## Testing

1.

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Python - local testing

- [ ] python 3.6
- [ ] python 3.7

### CODE REVIEW

**Comment by elijahbenizzy** [comment]
This s a WIP, but its on the right track. Things to do:

- [ ] Simplify the complexity around literal/upstream -- this might not be the best way to do it...
- [ ] Add a `@deprecated` decorator for decorators
- [ ] Delete the TODO exception
- [ ] Add `@parametrized_full` to docs
- [ ] Polish code
Then in a separate PR, after this is approved
- [ ] Refactor function_modifiers into separate files, and have an `__init__.py` that joins them
- [ ] Do the same with tests

**Comment by elijahbenizzy** [comment]
Need to run a ray/spark/dask test, but more likely will add some basic integration tests for everything -- E.G> a DAG that uses all our features.

File: decorators.md [hunk 1]
```diff
@@ -29,7 +64,11 @@ distinct outputs. The _parameter_ key word argument has to match one of the argu
...
 tuple(Output Name, Documentation string) -> value.
 
-## @parameterized_inputs
+Note that `@parametrized` is deprecated, and we intend for you to use `@parameterize_vales`. We're consolidating
+to make the parameterization decorators more consistent! But we will not break your workflow for a long time.
+
+## @parameterize_inputs (replacing @parameterized_inputs)
+
 Expands a single function into _n_, each of which corresponds to a function in which the parameters specified are mapped
 to the specified inputs. Note this decorator and `@parametrized` are quite similar, except that
```
Comment (skrawcz) [inline decorators.md:73]: update this line

File: hamilton/function_modifiers.py [hunk 2]
```diff
@@ -37,115 +41,211 @@ def get_default_tags(fn: Callable) -> Dict[str, str]:
...
+        for assigned_output, mapping in self.parametrization.items():
+            for parameter, value in mapping.items():
+                if not isinstance(value, ParametrizedDependency):
+                    bad_values.append(value)
+        if bad_values:
+            raise InvalidDecoratorException(f'@parameterize must specify a dependency type -- either upstream() or literal().'
+                                            f'The following are not allowed: {bad_values}.')
+        self.specified_docstrings = {key: value[1] for key, value in parametrization.items() if isinstance(value, tuple)}
 
     def expand_node(self, node_: node.Node, config: Dict[str, Any], fn: Callable) -> Collection[node.Node]:
```
Comment (skrawcz) [inline hamilton/function_modifiers.py:123]: can we break this function up at all into smaller testable chunks?

File: tests/test_dev_utils.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+
+@pytest.mark.parametrize(
+    'version_1, version_2, op',
+    [
+        (Version(0, 1, 2), Version(0, 1, 2), operator.eq),
+        (Version(0, 1, 2), Version(0, 1, 3), operator.lt),
+        (Version(0, 1, 2), Version(0, 1, 1), operator.gt),
+        (Version(1, 9, 0), Version(2, 0, 0), operator.lt),
+        (Version(2, 0, 0), Version(1, 9, 0), operator.gt),
```
Comment (skrawcz) [inline tests/test_dev_utils.py:15]: need to add (1, 11,0,'0rc') to the test for the if statement.

File: hamilton/function_modifiers.py [hunk 4]
```diff
@@ -37,115 +41,211 @@ def get_default_tags(fn: Callable) -> Dict[str, str]:
...
+                    '[name, doc string] -> value to function.')
+        super(parameterize_values, self).__init__(**{output: ({parameter: literal(value)}, documentation) for (output, documentation), value in assigned_output.items()})
+
+
+@deprecation.deprecated(
+    warn_starting=(1, 10, 0),
+    fail_starting=(2, 0, 0),
+    use_this=parameterize_values,
+    explanation='We now support three parametrize decorators. @parameterize, @parameterize_values, and @parameterize_inputs',
+    migration_guide='https://github.com/stitchfix/hamilton/blob/main/decorators.md'
```
Comment (skrawcz) [inline hamilton/function_modifiers.py:242]: we should probably have an explicit section to link to.

File: decorators.md [hunk 5]
```diff
@@ -4,7 +4,42 @@ While the 1:1 mapping of column -> function implementation is powerful, we've im
 business-logic reuse. The decorators we've defined are as follows
 (source can be found in [function_modifiers](hamilton/function_modifiers.py)):
 
-## @parameterized
+## @parameterize
+´xpands a signle function into n, each of which correspond to a function in which the parameter value is replaced either by:
```
Comment (skrawcz) [inline decorators.md:8]:
```suggestion
Expands a single function into n, each of which correspond to a function in which the parameter value is replaced either by:
```

Comment (skrawcz) [reply decorators.md:8]: also should make an explicit migration section that we can link to.

File: decorators.md [hunk 6]
```diff
@@ -68,9 +107,13 @@ def SOME_OUTPUT_NAME(SOME_INPUT_NAME: pd.Series) -> pd.Series:
     """SOME_INPUT_NAME shifted by 1 to create SOME_OUTPUT_NAME"""
     return SOME_INPUT_NAME.shift(1)
 ```
+Note that `@parameterized_inputs` is deprecated, and we intend for you to use `@parameterize_inputs`. We're consolidating
+to make the parameterization decorators more consistent! But we will not break your workflow for a long time.
```
Comment (skrawcz) [inline decorators.md:111]:
```suggestion
to make the parameterization decorators more consistent! You have plenty of time to migrate, we wont make this a hard change until we have a Hamilton 2.0.0 to release.
```

**Review by skrawcz — COMMENTED** [review]
I think we should have a clearer migration section that we can link to so people know what to do and until when.

File: tests/test_dev_utils.py (hunk 3 above)
Comment (elijahbenizzy) [reply tests/test_dev_utils.py:15]: Its not allowed in this case but I'll test the parser that strips it

File: hamilton/function_modifiers.py (hunk 2 above)
Comment (elijahbenizzy) [reply hamilton/function_modifiers.py:123]: Don't really think its worthwhile? Its about 40 lines, which doesn't seem ridiculous to me, and all kind of does one thing (gathers the nodes). *Might* be able to do something aroudn the docstirng, but I 've already streamlined that to some extent.

**Pull request merged**